### DISCUSSION TITLE
User mode (`--user`)

### DISCUSSION THREAD

**Discussion by GioF71** [GioF71/mpd-alsa-docker#372] [Announcements]
Hello, starting 2023-12-19, we have initial support for running the container in docker user mode.
At the moment, it is supported without pulseaudio outputs. For those it will be needed to provide some files as mount points, and it's something for another day.
Here is a working example with alsa output:

```text
---
version: '3.3'

services:
  mpd-d10-vanilla:
    image: giof71/mpd-alsa:latest
    container_name: mpd-d10-vanilla
    user: "1000:29"
    devices:
      - /dev/snd:/dev/snd
    networks:
      - mpd
    ports:
      - 6602:6600
    environment:
      - TZ=Europe/Rome
      - ALSA_OUTPUT_CREATE=yes
      - ALSA_OUTPUT_PRESET=topping-d10
      - ALSA_OUTPUT_NAME=d10-vanilla
    volumes:
      - /var/cache/docker-data/mpd/d10-vanilla/db:/db
      - /var/cache/docker-data/mpd/d10-vanilla/log:/log
      - ../../mpd/radio:/playlists
    restart: always
```

Make sure that the path you mount for /db and /log are writable by the uid/gid of your choice.
The group id should be the host 'audio' group if you want to use alsa outputs.

If the path /db and /log are not writable, the container will write to /tmp/db and /tmp/log respectively.

**Comment by GioF71** [comment]
And pulse audio version:

```text
---
version: "3"

services:
  mpd:
    image: giof71/mpd-alsa:vanilla
    container_name: mpd-pulse
    user: 1000:1000
    ports:
      - 6600:6600
    environment:
      - PULSE_AUDIO_OUTPUT_CREATE=yes
    volumes:
      - /run/user/1000/pulse:/run/user/1000/pulse
      - ./pulse-client.conf:/etc/pulse/client.conf:ro
      - ./db:/db
      - /.log:/log
```

Contents of pulse-client.conf:

```text
# Connect to the host's server using the mounted UNIX socket
default-server = unix:/run/user/1000/pulse/native

# Prevent a server running in the container
autospawn = no
daemon-binary = /bin/true

# Prevent the use of shared memory
enable-shm = false
```

Notice there is no `restart: unless-stopped` or similar. It is better to run this container with a user-level systemd service if you want automatic startup.

Make sure that the path you mount for /db and /log are writable by the uid/gid of your choice.
The group id should be the host 'audio' group if you want to use alsa outputs.

If the path /db and /log are not writable, the container will write to /tmp/db and /tmp/log respectively.

**Comment by GioF71** [comment]
Gists for pulse-client.conf for [uid=1000](https://gist.github.com/GioF71/7a7a2a3d5b0cc1552dfd84721d9e1edd) and [uid=1001](https://gist.github.com/GioF71/e168466753fc3b1e1092d926b8fb69a4)

**Comment by iconoclasthero** [comment]
I read over the readme:
> User mode is now possible when using alsa outputs, and of course it is mandatory (enforced) when using pulse outputs.

I'm not sure what the words 'mandatory' and 'enforced' are in this context. Running systemd --system mpd to pipewire-pulse is quite straightforward. 

Create a pulse output, e.g.:
```
audio_output {
    name           " Pulse Output to localhost "
    enabled        "no"
    type           "pulse"
    server         "127.0.0.1"
    scale_volume   "1.5"
#   sink           "remote_server_sink"    # optional
}
```
Use the pipewire-pulse native tcp module to go get the stream:
```
pactl unload-module module-native-protocol-tcp 2>/dev/null
pactl load-module module-native-protocol-tcp
```

The same applies to another machine on the network, e.g., my laptop at 192.168.1.3:
```
audio_output {
    name         " Pulse Output to Laptop "
    enabled      "no"
    type         "pulse"
    server       "192.168.1.3"           # optional
    scale_volume "1.0"
#   sink         "remote_server_sink"    # optional
}
```

So since both of these work with bare-metal mpd, I would think that it would be trivial to do with a docker, no?

Various information is available about how to secure the connection, [e.g.](https://gist.github.com/xarinatan/c415341ff34eab445cfb073988dcf6c1), but I've never been clear how this is relevant inside a LAN I control—and if someone else is listening to my mpd instance, great.

**Comment by GioF71** [comment]
Hello, about the "mandatory" and "enforced" terms, let me try to clarify:

- mandatory: When I implemented the pulseaudio feature, I took inspiration from [this work](https://github.com/TheBiggerGuy/docker-pulseaudio-example) (and hopefully I did not forget to give credit somewhere in the repository), and I thought that running as an user was mandatory. However, in the last few weeks, an user mentioned a different use case where he wants to use pulseaudio in a headless system, see [here](https://github.com/GioF71/upmpdcli-docker/issues/312#issuecomment-2323473865), so I would like to try if I can avoid this hard requirement. But, currently, running as a specific user_id and group_id (from PUID and PGID) is "mandatory" when using a PulseAudio output.

- enforced: the startup script, if a pulseaudio output is specified using the relevant env variables, enforces user mode using default values (1000) for PUID and PGID, so, if you specify a pulseaudio output, the container will run as 1000:1000 unless the user specifies PUID and PGID.

If these are not the most appropriate terms, please suggest, and I will gladly update the documentation.

About configuration, it seems to me that all the options are covered by the variables in [this section](https://github.com/GioF71/mpd-alsa-docker#pulseaudio-additional-outputs).
If that is not the case and/or you need some currently "unsupported" features, you can specify additional outputs on a user-provided file, see [here](https://github.com/GioF71/mpd-alsa-docker?tab=readme-ov-file#user-configuration-volume).
If something is missing, just notify me, and I will add it, if possible.

The other things you mention, like loading modules, should be done on the host system, or am I wrong?

Let me know if this helps!
